File: bench.py
import os
import time
import timeit
from typing import Dict
from multiprocessing import Process, Queue

# our contestants:
import cbrrr     # https://github.com/DavidBuchanan314/dag-cbrrr
import libipld   # https://github.com/MarshalX/python-libipld
import dag_cbor  # https://github.com/hashberg-io/dag-cbor

decoders = {
	"cbrrr": cbrrr.decode_dag_cbor,
	"libipld": libipld.decode_dag_cbor,
	"dag_cbor": dag_cbor.decode,
}

encoders = {
	"cbrrr": cbrrr.encode_dag_cbor,
	# libipld does not support encoding
	"dag_cbor": dag_cbor.encode,
}

DATA_DIR = "./data/"

testcases: Dict[str, bytes] = {}

for testcase in os.listdir(DATA_DIR):
	data = open(DATA_DIR + testcase, "rb").read()
	testcases[testcase] = data

print()
print("Hello World Decode:")
print("===================")
HELLO_ITERS = 100000
for name, decodefn in decoders.items():
	res = timeit.timeit(lambda: decodefn(testcases["trivial_helloworld.dagcbor"]), number=HELLO_ITERS)
	ns_per_it = (res / HELLO_ITERS) * 1000 * 1000 * 1000
	print(f"{name:<9}: {ns_per_it:.0f} ns")

print()
print("Hello World Encode:")
print("===================")
HELLO_ITERS = 100000
for name, encodefn in encoders.items():
	decoded = decoders[name](testcases["trivial_helloworld.dagcbor"])
	res = timeit.timeit(lambda: encodefn(decoded), number=HELLO_ITERS)
	ns_per_it = (res / HELLO_ITERS) * 1000 * 1000 * 1000
	print(f"{name:<9}: {ns_per_it:.0f} ns")

print()
print("Realistic Decode Tests:")
print("=======================")
REALISTIC_ITERS = 25
for testname, testcase in testcases.items():
	if testname.startswith("torture_") or testname.startswith("trivial_"):
		continue
	for decodername, decodefn in decoders.items():
		res = timeit.timeit(lambda: decodefn(testcase), number=REALISTIC_ITERS)
		res /= REALISTIC_ITERS
		ms_per_it = res * 1000
		mbps = len(testcase)/1024/1024/res
		print(f"{testname:<30} {decodername:<9}: {ms_per_it:.2f} ms ({mbps:.2f} MB/s)")

print()
print("Realistic Encode Tests:")
print("=======================")
REALISTIC_ITERS = 25
for testname, testcase in testcases.items():
	if testname.startswith("torture_") or testname.startswith("trivial_"):
		continue
	for encodername, encodefn in encoders.items():
		decoded = decoders[encodername](testcase)
		res = timeit.timeit(lambda: encodefn(decoded), number=REALISTIC_ITERS)
		res /= REALISTIC_ITERS
		ms_per_it = res * 1000
		mbps = len(testcase)/1024/1024/res
		print(f"{testname:<30} {encodername:<9}: {ms_per_it:.2f} ms ({mbps:.2f} MB/s)")

print()
print("Decode Torture Tests:")
print("=====================")
for testname, testcase in testcases.items():
	if not testname.startswith("torture_"):
		continue
	for decodername, decodefn in decoders.items():
		q = Queue()
		def run_and_except():
			try:
				decodefn(testcase)
				q.put("success")
			except Exception as e:
				q.put(str(e))
		p = Process(target=run_and_except)
		start = time.time()
		p.start()
		p.join()
		duration = time.time() - start
		if p.exitcode:
			print(testname.ljust(30, " "), decodername.ljust(9, " "), "SEGFAULT" if p.exitcode == -11 else "ERROR")
			continue

		res = q.get()
		if res == "success":
			duration_ms = duration * 1000
			mbps = len(testcase)/1024/1024/duration
			print(testname.ljust(30, " "), decodername.ljust(9, " "), f"{duration_ms:.1f} ms ({mbps:.2f} MB/s)")
		else:
			print(testname.ljust(30, " "), decodername.ljust(9, " "), "ERROR:", res)

File: ruby/bench.rb
#!/usr/bin/env ruby

require 'benchmark'

require 'cbor'  # https://github.com/cabo/cbor-ruby

DECODERS = {
  "cbor-ruby" => CBOR
}

ENCODERS = {
  "cbor-ruby" => CBOR
}

DATA_DIR = File.expand_path("../data", __dir__)

TESTCASES = Hash[Dir["#{DATA_DIR}/*.dagcbor"].sort.map { |f|
  [File.basename(f, '.dagcbor'), File.read(f, encoding: 'BINARY')]
}]

def run_test(title, mode, cases, iters: 1, unit: 'ns', show_mbps: false)
  puts
  puts "#{title}:"
  puts "=" * (title.length + 1)

  if cases.is_a?(String)
    cases = [cases]
    show_file_name = false
  elsif cases.is_a?(Array)
    show_file_name = true
  else
    raise "cases parameter should be an array"
  end

  method = mode

  cases.each do |test_name|
    data = TESTCASES[test_name]
    raise "Test not found: #{test_name.inspect}" unless data

    coders = (mode == :decode) ? DECODERS : ENCODERS

    coders.each do |lib_name, coder|
      test_label = show_file_name ? test_name.ljust(31) : ''
      lib_label = lib_name.ljust(coders.keys.map(&:length).max)

      begin
        if mode == :decode
          # give each impl a "clean slate", for maximum fairness
          GC.start

          time = Benchmark.realtime {
            iters.times { coder.decode(data) }
          }
        else
          decoded = DECODERS[lib_name].decode(data)
          GC.start

          time = Benchmark.realtime {
            iters.times { coder.encode(decoded) }
          }
        end

        time_per_iter = time / iters
        time_precision = (unit == 'ms') ? 2 : 0
        time_in_unit = time_per_iter * (10 ** (unit == 'ms' ? 3 : 9))
        formatted_time = sprintf("%.#{time_precision}f", time_in_unit) + ' ' + unit

        if show_mbps
          mbps = data.length * 1.0 / (1024 * 1024) / time_per_iter
          mb_speed = "(#{sprintf('%.2f', mbps)} MB/s)"
        else
          mb_speed = ""
        end

        puts "#{test_label}#{lib_label} : #{formatted_time} #{mb_speed}"
      rescue Exception => e
        puts "#{test_label}#{lib_label} : #{e.inspect}"
      end
    end
  end
end


HELLO_ITERS = 2_000_000

run_test("Hello World Decode", :decode, 'trivial_helloworld', iters: HELLO_ITERS)
run_test("Hello World Encode", :encode, 'trivial_helloworld', iters: HELLO_ITERS)

REALISTIC_ITERS = 100

realistic_tests = TESTCASES.keys.reject { |k| k.start_with?("torture_") || k.start_with?("trivial_") }

run_test("Realistic Decode Tests", :decode, realistic_tests, iters: REALISTIC_ITERS, unit: 'ms', show_mbps: true)
run_test("Realistic Encode Tests", :encode, realistic_tests, iters: REALISTIC_ITERS, unit: 'ms', show_mbps: true)

torture_tests = TESTCASES.keys.select { |k| k.start_with?("torture_") }

run_test("Decode Torture Tests", :decode, torture_tests, unit: 'ms', show_mbps: true)
